src/full/Agda/TypeChecking/Monad/Pure.hs

```diff
@@ -43,3 +45,9 @@ instance PureTCM m => PureTCM (MaybeT m)
 instance PureTCM m => PureTCM (ReaderT r m)
 instance (PureTCM m, Monoid w) => PureTCM (WriterT w m)
 instance PureTCM m => PureTCM (StateT s m)
+
+unKleisliTCM :: forall m a b. PureTCM m => (forall n. PureTCM n => a -> n b) -> m (a -> b)
+unKleisliTCM k = do
+  state <- getTCState
+  env <- askTC
+  pure $ \a -> unsafePerformIO (fst <$> runTCM env state (k a))
```

Comment:
This implementation is perhaps a bit evil, but since `unSpine` is used in pure contexts, I thought that this would be easier than turning `unSpine'` into a monadic function. Happy to make that change if we think this is _too_ evil.

Reply:
`unsafePerformIO` is a heavy weapon...  It is not clear to me when a use of `unKleisliTCM` is sound.

The polymorphic type here might also sabotage @AndrasKovacs ' efforts of giving GHC more opportunities to generate efficient code.

Maybe you can instead drill open the `getDisplayForms` function into a `displayFormGetter :: m (QName -> [LocalDisplayForms])` ? https://github.com/agda/agda/blob/561fdc2f87e732cc8ea8c3a2223041efef3a63d3/src/full/Agda/TypeChecking/Monad/Signature.hs#L738-L744

Reply:
It's pretty annoying to make a "factory" version of `getDisplayForms` because `getConstInfo` (even `defaultGetConstInfo`) still wants to be in a monad with debug info — and I really don't want to refactor `unSpine'` since we'd be at GHC's mercy to SPECIALIZE it for the identity. Fortunately we already have a less nasty `unKleisli` laying around — in fast reduce — I just didn't realise that `MonadReify` implied `MonadReduce`! Patch incoming..

Reply:
> The polymorphic type here might also sabotage @AndrasKovacs ' efforts of giving GHC more opportunities to generate efficient code.

Also, you're right, of course :woman_facepalming: I was hoping that GHC would be able to pass `unKleisliTCM` the optimised version of `hasDisplayForms`, but I seem to have temporarily forgotten how dictionary arguments work. Sorry, this function is terrible on all fronts!

Reply:
If  `unKleisliTCM` is `INLINE`, the use sites are still fairly likely to be fully specialized.

Reply:
If `unKleisliTCM` is inlined then I'm confident that its use site will be fully specialized.